### ISSUE TITLE
VM: GC exhibits O(N^2) time in slowly growing a huge heap

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#25954]
A program that does a lot of allocations and keeps ~10% of them experiences increasing proportion of GC time as the heap grows. This 'retain 10%' is reasonable rough model for the execution of __dart2js__. Large dart2js compilations are now nearly 30% GC time.

There is a limit, __--old_gen_growth_rate=280__, that forces arithmetic expansion of old_gen heap.
The arithmetic expansion causes GC time to scale as the square of the heap size. The GC policy should avoid the need for tweaking manual GC parameters by incorporating a geometric aspect to avoid quadratic GC time.

One idea is that the heap growth is limited to the __max__ of __old_gen_growth_rate__ and a factor similar to __--old_gen_space_ratio=20__. Then the heap will grow by at most by, say, 20% beyond the live data aftr GC, but will grow smaller heaps by more rapidly.

The following program demonstrates the quadratic behaviour.

```sh
# parameter is GB of allocation.
$ sdk/bin/dart gc2.dart 1
permanent = 44739242  ephemeral = 671088620  30.369s
$ sdk/bin/dart gc2.dart 2
permanent = 89478485  ephemeral = 1342177280  102.468s
$ sdk/bin/dart gc2.dart 3
permanent = 134217728  ephemeral = 2013265920  220.894s
```

gc2.dart:
```dart
class Node {
  Node left;
  Node right;
  static const int size = 3 * 8; // header + 2 fields
}

var permanentRoot = [];
int permanentTick = 0;

main(List<String> arguments) {
  int b1 = 0;
  int b2 = 0;
  // Number of nodes required to allocate <argument>GB
  int target = (double.parse(arguments[0]) * (1024*1024*1024)) ~/ Node.size;
  int permanent = 0;
  int ephemeral = 0;
  Node ephemeralRoot;
  Stopwatch sw = new Stopwatch()..start();

  while (permanent < target) {

    b1 = 255 & (b1 + 1);

    b2 = 0xFFFFFF & (b2 + (b1 & 1) + 1); // +3 every 2 iterations

    var node = new Node();

    if ((b2 & (15 << 1)) == 0) {  // approx 3/(16*2) = ~10% of allocations.
      if (permanent % 100 == 0) permanentRoot.add(null);
      ++permanent;
      var prev = permanentRoot[permanentTick];
      permanentRoot[permanentTick] = node;
      node.left = prev;
      permanentTick = (permanentTick + 1) % permanentRoot.length;
    } else {
      ++ephemeral;
      if (ephemeralRoot == null) {
        ephemeralRoot = node;
      } else {
        node.left = ephemeralRoot;
        ephemeralRoot = node;
      }
    }
    if ((b2 & (255 << 2)) == 0) {
      ephemeralRoot = null;
    }
  }
  print('permanent = ${permanent}  ephemeral = ${ephemeral}'
        '  ${sw.elapsed.inMilliseconds / 1000}s');
}
```

**Comment by iposva-google** [comment]
cc @a-siva

**Comment by a-siva** [comment]
https://github.com/dart-lang/sdk/commit/85cb1b3bd82f8f860e9a5aba15796f7c727b5a81

**Comment by rakudrama** [comment]
I'm excited to see this work but I'm re-opening the issue because the changes result in only a modest improvement for dart2js.

I compiled a large app. I pressed the 'Refresh' button on the home page as soon as the isolate exited and took numbers from the pie chart.

Before: 24.98% GCOldSpace  6.08% GCNewSpace
After:  18.94% GCOldSpace 6.66% GCNewSpace

I.e. total GC reduced from 31.06% to 26.6%.

If I run with --old-gen-growth-rate=1000, I get something that is clearly better:
7.88% GCOldSpace 7.66% GCNewSpace

At --old-gen-growth-rate=2000: 4.21% GCOldSpace 7.81% GCNewSpace

I'd like to see the automatic policy get closer to these manual results, say, always <15% GC.
